ddtrace/contrib/aiohttp/middlewares.py

```diff
@@ -1,21 +1,76 @@
-import asyncio
+import functools
 
 from ..asyncio import context_provider
 from ...compat import stringify
 from ...constants import ANALYTICS_SAMPLE_RATE_KEY, SPAN_MEASURED_KEY
 from ...ext import SpanTypes, http
 from ...propagation.http import HTTPPropagator
 from ...settings import config
+from ...context import Context
 
+try:
+    from aiohttp.web import middleware
 
-CONFIG_KEY = 'datadog_trace'
-REQUEST_CONTEXT_KEY = 'datadog_context'
-REQUEST_CONFIG_KEY = '__datadog_trace_config'
-REQUEST_SPAN_KEY = '__datadog_request_span'
+    AIOHTTP_HAS_MIDDLEWARE = True
+except ImportError:
+    AIOHTTP_HAS_MIDDLEWARE = False
 
+    def middleware(f):
+        return f
 
-@asyncio.coroutine
-def trace_middleware(app, handler):
+
+CONFIG_KEY = "datadog_trace"
+REQUEST_CONTEXT_KEY = "datadog_context"
+REQUEST_CONFIG_KEY = "__datadog_trace_config"
+REQUEST_SPAN_KEY = "__datadog_request_span"
+
+
+@middleware
+async def trace_middleware_2x(request, handler, app=None):
+    # application configs
+    if app is None:
+        app = request.app
+
+    tracer = app[CONFIG_KEY]["tracer"]
+    service = app[CONFIG_KEY]["service"]
+    distributed_tracing = app[CONFIG_KEY]["distributed_tracing_enabled"]
+
+    # Create a new context based on the propagated information.
+    if distributed_tracing:
+        propagator = HTTPPropagator()
+        context = propagator.extract(request.headers)
+
+        if context.trace_id:
+            # activate propagated context
+            tracer.context_provider.activate(context)
+        else:
+            # clear out any existing context for new request
+            tracer.context_provider.activate(Context())
```

Comment:
updated description to state:
```python
            # In case a non-distributed request comes after a distributed request we need to clear out
            # the previous context
```

Reply:
@thehesiod This should be handled by the ContextVars-enabled default context manager:

https://github.com/DataDog/dd-trace-py/blob/master/ddtrace/internal/context_manager.py#L92-L98